### PR TITLE
Add netstandard2.1 build targets & use System.Security.Cryptography.AesCcm when available

### PR DESCRIPTION
**Pull request by michaelgayeski**
Hello, this change uses System.Security.Cryptography.AesCcm class to encrypt and decrypt SMB messages on supported platforms. This performs 4-9x faster than the current encryption scheme and is more resource-efficient. If the scheme is not supported it uses the implementation found in the Utilities library.

### CODE REVIEW

**Comment by michaelgayeski** [comment]
image

Here's a bench

File: SMBLibrary.Tests/SMB2EncryptionTests.cs [hunk 1]
```diff
@@ -7,11 +7,12 @@
 using System;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 using SMBLibrary.SMB2;
+using SMBLibrary.SMB2.Encryption;
 using Utilities;
 
 namespace SMBLibrary.Tests
 {
-    [TestClass]
+	[TestClass]
```
Comment (TalAloni) [inline SMBLibrary.Tests/SMB2EncryptionTests.cs:15]: Why did you change the indentation here?

File: SMBLibrary/Client/SMB2Client.cs [hunk 2]
```diff
@@ -12,11 +12,12 @@
 using System.Threading;
 using SMBLibrary.NetBios;
 using SMBLibrary.SMB2;
+using SMBLibrary.SMB2.Encryption;
 using Utilities;
 
 namespace SMBLibrary.Client
 {
-    public class SMB2Client : ISMBClient
+	public class SMB2Client : ISMBClient, IDisposable
```
Comment (TalAloni) [inline SMBLibrary/Client/SMB2Client.cs:20]: Why did you change the indentation

File: SMBLibrary.Tests/SMB2EncryptionTests.cs [hunk 3]
```diff
@@ -69,10 +70,20 @@ public void TestEncryption()
...
+            DefaultAesCcmEncryptionProvider encryptionStrategy = new DefaultAesCcmEncryptionProvider(encryptionKey);
+
+            SMB2TransformHeader transformHeader = SMB2Cryptography.CreateTransformHeader(nonce, message.Length, sessionID);
+
+            byte[] encryptedMessage = encryptionStrategy.EncryptMessage(nonce, message, transformHeader.GetAssociatedData(), out signature);
 
             Assert.IsTrue(ByteUtils.AreByteArraysEqual(expectedEncrypted, encryptedMessage));
             // The associated data in this sample include non-zero nonce padding so we ignore signature validation
+
+#if NETCOREAPP3_1_OR_GREATER
```
Comment (TalAloni) [inline SMBLibrary.Tests/SMB2EncryptionTests.cs:82]: IMO the test should run once under each supported platforms, no need to test encryption twice

File: SMBLibrary/Client/SMB2Client.cs [hunk 4]
```diff
@@ -246,8 +246,21 @@ public NTStatus Login(string domainName, string userName, string password, Authe
                             if (m_dialect == SMB2Dialect.SMB300)
                             {
                                 m_encryptSessionData = (((SessionSetupResponse)response).SessionFlags & SessionFlags.EncryptData) > 0;
-                                m_encryptionKey = SMB2Cryptography.GenerateClientEncryptionKey(m_sessionKey, SMB2Dialect.SMB300, null);
-                                m_decryptionKey = SMB2Cryptography.GenerateClientDecryptionKey(m_sessionKey, SMB2Dialect.SMB300, null);
+
+                                if (m_encryptionProvider != null)
+								{
```
Comment (TalAloni) [inline SMBLibrary/Client/SMB2Client.cs:251]: The indentation here is inconsistent with other places

File: SMBLibrary/Client/SMB2Client.cs [hunk 5]
```diff
@@ -415,7 +428,7 @@ private void ProcessPacket(SessionPacket packet, ConnectionState state)
                 {
                     SMB2TransformHeader transformHeader = new SMB2TransformHeader(packet.Trailer, 0);
                     byte[] encryptedMessage = ByteReader.ReadBytes(packet.Trailer, SMB2TransformHeader.Length, (int)transformHeader.OriginalMessageSize);
-                    messageBytes = SMB2Cryptography.DecryptMessage(m_decryptionKey, transformHeader, encryptedMessage);
+                    messageBytes = SMB2Cryptography.DecryptMessage(m_decryptionProvider, transformHeader, encryptedMessage);
```
Comment (TalAloni) [inline SMBLibrary/Client/SMB2Client.cs:431]: I understand that we want to use different provider, what I don't understand is changing the DecryptMessage method signature

File: SMBLibrary/Client/SMB2Client.cs [hunk 6]
```diff
@@ -619,13 +633,13 @@ public uint MaxWriteSize
             }
         }
 
-        public static void TrySendCommand(Socket socket, SMB2Command request, byte[] encryptionKey)
+        public void TrySendCommand(Socket socket, SMB2Command request, bool encrypt)
```
Comment (TalAloni) [inline SMBLibrary/Client/SMB2Client.cs:636]: Again, why is it necessary to change the method signature?

File: SMBLibrary/Client/SMB2Client.cs [hunk 7]
```diff
@@ -584,7 +597,8 @@ internal void TrySendCommand(SMB2Command request, bool encryptData)
                     request.Header.Signature = ByteReader.ReadBytes(signature, 0, 16);
                 }
             }
-            TrySendCommand(m_clientSocket, request, encryptData ? m_encryptionKey : null);
+
```
Comment (TalAloni) [inline SMBLibrary/Client/SMB2Client.cs:600]: I don't understand why you changed the styling here

**Comment by TalAloni** [comment]
Hi Michael,
While I appreciate the sentiment behind the PR, and the work you did, the PR itself does a lot of things other than using AesCcm.
When I create PRs for other projects, I strive that each PR would focus on a single goal.
Your pull request make multiple changes, and I'm not convinced that all of them are necessary to capture on the AesCcm optimization.
Changing method signatures, inheriting from IDisposable and changing coding styles seems like optional changes that may be avoided or achieved in a separate PR.

**Comment by michaelgayeski** [comment]
Thank you for the feedback, @TalAloni, I'll break this apart, revert the style changes; remove anything absolutely unnecessary, and have this one focus soley on the AES change.

**Comment by TalAloni** [comment]
Also, while it's nice that AesCcm provides improved performance, does it really make an impact within the wider context of SMB? when you factor in network latency - does it really make a difference to throughput?

**Comment by michaelgayeski** [comment]
> Also, while it's nice that AesCcm provides improved performance, does it really make an impact within the wider context of SMB? when you factor in network latency - does it really make a difference in terms of throughput?

Yes, absolutely; especially when doing large transfers on machines with limited CPU. I can publish a bench of that scenario if you'd like.  For smaller or more infrequent transfers it's probably not as significant, but still more efficient.

File: SMBLibrary.Tests/SMB2EncryptionTests.cs (hunk 1 above)
Comment (michaelgayeski) [reply SMBLibrary.Tests/SMB2EncryptionTests.cs:15]: Thanks, looks like my tab/spacing settings are all messed up. I will revert these.

File: SMBLibrary/Client/SMB2Client.cs (hunk 5 above)
Comment (michaelgayeski) [reply SMBLibrary/Client/SMB2Client.cs:431]: Sure, there's a bit more going on in the DecryptMessage method (extracting the headers, associated data, etc..) that I was trying not to duplicate. 

I wanted to use the same instance of the AesCcm class for a given key as creating/tearing them down involves a bit of overhead, and to do that needed to pass in that additional state. Since this was an internal class I thought It might be more appropriate to change this signature to and remove the selection of the actual cryptographic implementation from this method. 

I can do this differently if you think it is more appropriate.

**Comment by TalAloni** [comment]
I apologize - I'm quite busy these days and haven't got around to look into this change yet.

**Comment by michaelgayeski** [comment]
@TalAloni Thanks, me as well. I am going to do a bit more cleanup per your feedback. I will close it for now until it is ready so it is off your radar :)
